### ISSUE TITLE
How to generate CREATE TABLE IF NOT EXISTS from prisma schema?

### ISSUE THREAD

**Issue by mumincelal** [prisma/prisma#13068]
When I write a model into the schema.prisma file, generated migration SQL is becoming `CREATE TABLE <table_name>` but I want generated SQL to become `CREATE TABLE IF NOT EXISTS <table>`.

Generated SQL file can be edited but it is better to be handled by Prisma.

**Comment by tomhoule** [comment]
Hi @mumincelal — it's not possible to write your schema in such a way that migrate will generate `CREATE TABLE IF NOT EXISTS`, and in general it is going to conflict with how migration generation (diffing) works in migrate: if the table is in the database, it will not generate a `CREATE TABLE`, otherwise it will generate a `DROP TABLE`. Can you elaborate on your specific use case?

**Comment by mumincelal** [comment]
Hi @tomhoule,

The reason for this request is to prevent dropping an existing table. In the current situation, when we try to migrate changes  to database, prisma will run always the `CREATE TABLE` command but that is already in database. So will this command fail?

**Comment by tomhoule** [comment]
Ah, if this is a new project, the solution you are looking for is [baselining](https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/baselining) through introspection — you can have one migration reflecting the start state of your main database in your development environment, then tell migrate not to apply that one on your main database through `prisma migrate resolve --applied`.

**Comment by janpio** [comment]
> In the current situation, when we try to migrate changes to database, prisma will run always the `CREATE TABLE` command but that is already in database.

Or asked another way: Why does your project create a migration to create a table if the table already exists in the database?

**Comment by GovindarajanNagarajan-TomTom** [comment]
> Or asked another way: Why does your project create a migration to create a table if the table already exists in the database?

Need this feature as well. We have projects where we are moving databases and tables to Prisma incrementally from an existing project. Ideally, Prisma should have some options to configure this so that we don't write over existing tables and allow us to incrementally add more management via Prisma. 

Other ORMs (for example SQLAlchemy allow this : https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.MetaData.create_all)

**Comment by janpio** [comment]
We are not going to add this based on the use case described in the original issue.

`IF NOT EXISTS` only would make sense if you already have a table present, and Prisma generates a migration that creates this table anyway. If that happens, you either should be using [baselining](https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/baselining) to _avoid_ that it does - or something went wrong and your database is already in a state where Prisma Migrate will struggle and not work properly.

If you have concrete use cases that depend on something like this, please open a new issue or discussion and describe that use case so we can take it apart and figure out what the correct usage of Prisma-related command (already existing or to be built) would be.

@GovindarajanNagarajan-TomTom Can you describe how that SQLAlchemy feature is used?

**Comment by dixonmyles** [comment]
To throw my hat into the ring here, as a relatively new Prisma user, I ran into a similar desire after one of our data guys had started making changes to the database after we'd already switched over to using Prisma and therefore had already done a baseline as mentioned prior. This resulted in our development db (just a local postgres instance launched via docker-compose) being out of sync with our production db. These were the steps that I had taken, and if this is not the recommended way to have approached this I'd love to learn what the proper approach is:

- Changes were made to the production db via a separate set of sql commands run directly against the DB
- I pulled the changes into my `schema.prisma` via `prisma db pull`, specifying the production db as my datasource
- I switched my datasource back to the development db and applied those changes to it with `prisma migrate dev`
- This generated a `migration.sql` file with a large number of `CREATE TABLE` and other similar commands
- When my CICD pipeline reached the `prisma migrate deploy` step, this new migrationfailed due to those tables and relations already existing in the production db
- As there had been no other changes to the `schema.prisma` file after doing the introspection, I opted to manually mark the migration as resolved in the production db via `prisma migrate resolve --applied <migration>`

The outcome to this is my production and dev db are now back in sync and both dbs show all migrations as resolved, and my `schema.prisma` file matches both. Is there a simpler solution that I could've taken? In the same vein, is it really expected/intended that changes are made to the db both via sql commands run directly against it and changes to the `schema.prisma` file?

**Comment by janpio** [comment]
Yes, these are roughly the steps - both your Prisma schema and your migration history need to be updated for Prisma in the future to be able to generate correct migrations, and apply them reliably.

You could also get the SQL command that the external party ran, add it to a new migration in your app instead of using `migrate dev` locally. But you still need to mark it as applied on the modified database via `migrate resolve` vs. actually applying it on databases that are still in the correct, Prisma-expected state.

(If Prisma used `CREATE TABE IF NOT EXISTS` this would get a lot messier, as the existing table structure does not have to match the one the migration file tried to create - and then it would only fail on a future migration or when trying to read or write data from these tables.)

**Comment by Domains18** [comment]
I am facing the same issue on v7. My migration is failing because that type already exists. MY reasoning; if we used `IF NOT EXISTS` it would skip creating trying to create the type, would it not ? is there a workaround because I have to edit the migration files everytime, not so far from using raw sql like old times
